system/security/plugins/jwtSecurity/jwtCache.cpp

```diff
@@ -337,32 +371,53 @@ unsigned int JWTUserInfo::_countScopePerms(const PermissionMap& scopePermMap) co
 
 //=============================================================================
 
-bool JWTUserCache::has(const std::string& userName) const
+bool JWTUserCache::has(const std::string& userName, time_t expireTime) const
 {
-    CriticalBlock block(crit);
-    return userPermMap.find(userName) != userPermMap.end();
+    CriticalBlock                       block(crit);
+    UserPermissionMap::const_iterator   foundIter = userPermMap.find(userName);
+
+    if (foundIter != userPermMap.end())
+    {
+        time_t  savedTime;
+
+        std::tie(savedTime, std::ignore) = foundIter->second;
+
+        return expireTime < savedTime;
+    }
+
+    return false;
 }
 
 JWTUserCache& JWTUserCache::set(const std::string& userName, std::shared_ptr<JWTUserInfo>& userInfo)
 {
     {
-        CriticalBlock block(crit);
+        CriticalBlock   block(crit);
+        UserInfoTuple   info = std::make_tuple(time(nullptr), userInfo);
 
-        userPermMap[userName] = userInfo;
+        userPermMap[userName] = info;
     }
 
     return *this;
 }
 
-const std::shared_ptr<JWTUserInfo> JWTUserCache::get(const std::string& userName) const
+const std::shared_ptr<JWTUserInfo> JWTUserCache::get(const std::string& userName, time_t expireTime) const
 {
     {
         CriticalBlock                       block(crit);
         UserPermissionMap::const_iterator   foundIter = userPermMap.find(userName);
 
         if (foundIter != userPermMap.end())
         {
-            return foundIter->second;
+            time_t                          savedTime;
+            std::shared_ptr<JWTUserInfo>    userInfo;
+            std::shared_ptr<JWTUserInfo>    nullUserInfo;
```

Comment:
Could this be declared as `static`, or even as a member? It seems wasteful to declare this unchanging object every time.

Reply:
I've removed that branch entirely by adding checks for returned nullptr values in the calling functions.